### ISSUE TITLE
KPS Chinese version

### ISSUE THREAD

**Issue by Inlife233** [RoanH/KeysPerSecond#55]
I made a Chinese version of this software (just modify the source code), can I publish it? Can I submit my modified source code to this repository for other Chinese players to use?

**Comment by RoanH** [comment]
It would be really cool to see translations of the interface into a other languages for sure. Modifying the source code directly to do this is however not really the way to do this as it'll make it really hard for me to maintain during future releases. So I would implement a proper system for translations.

If you've already modified the source code with translations then please submit a pull request to this repository and I'll build the translation system on top of that.

**Comment by RoanH** [comment]
@Inlife233 Are you still interested in this?

**Comment by TeusDev** [comment]
> @Inlife233 I'm still interested in this, but I do need at least one translation to make this worth my time. So if you're still willing to help out that would be great. If creating the pull request is an issue you can also email me the modified source code or send a download link.

i can help with PT-BR translation. I don't have a modified source code to share on the other hand XD

**Comment by RoanH** [comment]
> i can help with PT-BR translation. I don't have a modified source code to share on the other hand XD

@TeusDev Great! Just knowing that you would be able to help out once a proper system for translations is in place is already amazing. I'll put this on the roadmap and then I'll still have to see when I have enough free time to implement such a large system.

**Comment by soloopooo** [comment]
Hello RoanH I can help for Zh-CN(Simplified Chinese) and Zh-TW(Traditional Chinese) translation.

Please feel free to ask me for Chinese translation when you are ready!

**Comment by RoanH** [comment]
> Hello RoanH I can help for Zh-CN(Simplified Chinese) and Zh-TW(Traditional Chinese) translation.
> 
> Please feel free to ask me for Chinese translation when you are ready!

Hey! Thank your very much for the offer and apologies for the very late reply... I think that simplified and traditional Chinese would be great to have as some of the first supported languages and it also gives me a good reason to increase the priority of this feature.

**Comment by soloopooo** [comment]
> Hey! Thank your very much for the offer and apologies for the very late reply... I think that simplified and traditional Chinese would be great to have as some of the first supported languages and it also gives me a good reason to increase the priority of this feature.

Thank you! If you finished your translation template, please just feel free to mention me again :)

**Comment by RoanH** [comment]
Hi, I haven't looked into this very extensively yet since I'm first trying to clean up the codebase before adding a feature as large as this. Currently, I'm primarily considering to use Java's built in [i18n](https://docs.oracle.com/javase/8/docs/technotes/guides/intl/index.html) implementation. But localisation has a lot of caveats and this program was never built with localisation in mind, so there are a number of places that might be hard to localise at the moment. So not sure if you'd want to take on such a large task (side note, the complete list of open issues is [here](https://git.roanh.dev/roan/KeysPerSecond/-/issues)).

**Comment by efojug** [comment]
Hi, I accidentally discovered on another project that it might be possible to use weblate to quickly contribute translations to a project on a website!
https://hosted.weblate.org Hope this helps!

**Comment by soloopooo** [comment]
Hello RoanH,
Before jumping into interface translation, I want to tell you some common issues that could be happen:
  1. Font compatibility, especially for CJK (Chinese, Japanese, Korean) users. You may not know that the font,or character patterns for Mainland China, Taiwan, Hong Kong, Japan and Korea are NOT THE SAME. I had seen so many wrongy font used in some steam games. You can refer to some Open-Source CJK font in github to find out the difference for CJK fonts. For example, you can search `lxgw` on github and see his font repo.
  2. Codec compatibility. This is also for CJK users. When we are on English, ascii is enough, but on CJK, it's not enough, and could get an issue called mojibake. To avoid that, we should use UTF-8 as the default codec.

**Comment by RoanH** [comment]
Yep, I am (unfortunately) aware of many of the issues surrounding proper internationalisation support..., but thanks for the detailed warnings! Even then, I'm sure I'll get some things wrong for some languages the first time around... Encoding support should already be up to the required level though, since I already support UTF-8 for panel names, e.g.:
![afbeelding](https://github.com/RoanH/KeysPerSecond/assets/8530896/02435b26-c9f2-4293-9d3a-e8c032f6a0db)
Though actual font configuration isn't a thing yet (but planned). Currently everything is rendered with the `Dialog` font, which is a logical font that Java itself maps to a sensible physical font present in the OS.
